### DISCUSSION TITLE
GSoC 2024: Android/iOS Support

### DISCUSSION THREAD

**Discussion by DenserMeerkat** [foss42/apidash#406] [General]
## Project Description

Enhance the usability of API Dash by extending its compatibility to Android and iOS platforms, while introducing essential features such as environment variables management, request history, and collection runners with monitoring. By providing mobile compatibility, developers will seamlessly access API Dash features on their smartphones and tablets, enhancing productivity and workflow. The addition of environment variables, request history, and collection runners with monitoring significantly boosts API Dash's functionality, empowering developers with comprehensive testing capabilities and streamlined workflow management. Adapting the desktop version's features into a mobile interface optimized for limited screen space ensures that API Dash remains intuitive, user-friendly, and accessible on both Android and iOS platforms.

[GSoC Project page](https://summerofcode.withgoogle.com/programs/2024/projects/exbL7COY)

## 📆 Week 1								[ May 27 - June 02 ]
- Setup breakpoints and logic to branch views. 
   - Mobile
   - Tablet
   - Desktop
- Use [flex_color_scheme](https://pub.dev/packages/flex_color_scheme) to apply surface blends.
- Implement [inner_drawer](https://pub.dev/packages/inner_drawer) for mobile view with drawers as per design.
- Style drawers and animations on navigating between drawers.
- Create a mobile navigation system with required destinations.
- Implement navigation to Settings & About page.

## 📆 Week 2								[ June 03 - June 09 ]
- Fix existing desktop widgets to work in mobile view.
- Ensure widgets are accessible with mobile inputs.
- Start design & development of Environment variable management feature.
- Implement UI for adding/updating environment variables.

## 📆 Week 3								[ June 10 - June 16 ]
- Add models & providers to implement the Environment manager.
- Write utility and helper functions required for Environment variable management.
- Modify required fields with [multi_trigger_autocomplete](https://pub.dev/packages/multi_trigger_autocomplete) to show available environment variables suggestions.
- Add styles to highlight variables present in the field.
- Add popover on hovering/tapping the variable to show value.

## 📆 Week 4								[ June 17 - June 23 ]
- Start design & development of History of Requests feature.
- Develop the UI to view the history of requests.
- Add models and providers to manage the history of requests.
- Add options to search, filter and delete.

## 📆 Week 5								[ June 24 - June 30 ]
- Test features implemented so far.
- Write new unit tests.
- Write widget tests for new widgets added.
- Fix any bugs/issues found while testing.

## 📆 Week 6								[ July 01 - July 07 ]
- Continue testing and bug fixing.
- Develop a user interface option to initiate the check for the latest release.
- Implement the option to download the latest release if available.
- Prepare documentation for midterm evaluation.

	
## 📆 Week 7								[ July 12 - July 19 ]
- Start design & development of Collection Runner & Monitoring feature.
- Develop UI for creating collections and adding requests to the collection.
- Add functionality for users to define expected response code, content type for each request.
- Add asserts for the response
    - Length of keys
    - Name of keys
    - Length of array values
    - Expected value
    - And more...

## 📆 Week 8								[ July 20 - July 26 ]
- Write functions to run requests sequentially with customizable options and assert response formats.
- Test collection runner feature.
- Implement responsive layout to handle larger screens tablets/iPads.
- Test on various device widths.

## 📆 Week 9								[ July 27 - Aug 02 ]
- Test the application on iOS.
- Ensure the UI is accessible with limited navigation buttons on iOS.
- Implement any necessary changes required.
- Write unit and widget tests.

## 📆 Week 10							[ Aug 03 - Aug 09 ]
- Add native splash screens using the [flutter_native_splash](https://pub.dev/packages/flutter_native_splash) package.
- Handle permissions while app onboarding with [permission_handler](https://pub.dev/packages/permission_handler).
- Continue unit and widget tests.
- Start working on integration testing & end-to-end workflow testing.

## 📆 Week 11							[ Aug 10 - Aug 16 ]
- Continue with integration tests & workflow tests.
- Fix any bugs/issues found while testing.

## 📆 Week 12							[ Aug 17 - Aug 23 ]
- Test release mode builds in physical devices.
- Perform quality and performance analysis.
- Final review with mentors.
- Prepare documentation for final evaluation.

**Comment by ashitaprasad** [comment]
**Meeting Notes**
Date: 8 June 2024

Action items to be undertaken by @DenserMeerkat:
- [x] Address PR #227 comments for the merge
- [x] Create a new PR for responsive Desktop App which switches between the layout based on available width & not specific to OS
- [x] Address all issues discussed for the mobile version and create a new PR after PR in action item 1 is merged
- [x] Continue working on env manager code
- [x] Start working on history of requests design
- [x] Move new code to using hooks for internal widget state management as it is much cleaner and get rid of `initState` & `dispose`

**Reply by DenserMeerkat** [reply]
Issues discussed in mobile version

- [ ] Create folder name docs and document non intuitive features.
- [x] Show Request bottom Appbar over keyboard.
- [x] Redesign request response navigation.
- [ ] Add share button in View Code page.

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 12 June 2024

Action items to be undertaken by @DenserMeerkat 

- [x]  Migrate to native Drawer since existing drawer implementation has issues on resizing. #408 
- [x]  Complete Environment editor - Variables table with checkbox. #412 
- [x]  Add dropdown in Requests page to toggle between environments. #412 
- [x]  Implement Text highlighting and Popover details card on hovering variable in fields. #412

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 16 June 2024

Action items to be undertaken by @DenserMeerkat

 - [x] Make Topbar action buttons (Rename, Delete, Duplicate) tray height slimmer.
 - [x] Remove dropdown from Environment Editor pane.
 - [x] Remove secret variables UI part.
 - [x] Remove Divider under Global sidebar card and make card standout instead.
 - [x] Fix Environment dropdown width.
 - [x] Implement pipeline to recognize variables in fields and map the to values.
 - [x] Fields which have to work with env vars:
       - URL field
       - Params <key, value>
       - Headers <key, value>
 >  Body - formdata <key, value>, JSON  _[Will be worked on later]_

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 22 June 2024

Action items to be undertaken by @DenserMeerkat

 - [ ] Tryout different styling for global environment sidebarcard.
 - [X] Fix environment field selection exception. [Issue](https://github.com/iWOLFSTEIN/mention_tag_text_field/issues/5)
 - [ ] Figure out a way to make environment suggestions work with header typeahead.
 - [x] Add drawer close button in very narrow width.
 - [x] Move env textfields to common folder.

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 27 June 2024

Action items to be undertaken by @DenserMeerkat

 - [x] Fix Material color deprecations in new Flutter version. #417
 - [x] Fix existing tests and figure out testing responsive layouts of different widths. #417
 - [x] Move About page into a dialog. #420 
 - [x] Figure out how to load history of requests hive boxes optimally.
 - [x] Double parenthesis triggering env variables leads to incorrect mentions. #429

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 30 June 2024

 - Discussed on implementation of History of Requests.
 - Reviewed all pending action items from previous meets.

**Comment by ashitaprasad** [comment]
Fixed #423

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 07 July 2024

Action items to be undertaken by @DenserMeerkat

 - [x] Fix Multi split view issue with resizing. 
    - [related issue](https://github.com/caduandrade/multi_split_view/issues/76)
    - #426
 - [x] Fix environment field copy paste with mentions issue. 
    - [raised issue](https://github.com/iWOLFSTEIN/mention_tag_text_field/issues/9)
    - #429 
 - [x] Add test for About dialog.
 - [x] Continue working on History of Requests.

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 10 July 2024

 - Discussed on choosing suitable fix for issue with resizing splitviews of the multi_split_view package.
 - Discussed on fixes for environment field issues.
   - **Copy/Paste issue**:
      implement extended_text_field to overcome copy paste issue.
   - **Double parenthesis triggering issue**:
      - implement a custom fork of multi_trigger_autocomplete to handle substring triggers.
      - handle custom triggerEnd instead of white space.
 - Continue with history of requests feature.
> custom fork of multi_trigger_autocomplete can be later added to foss42.

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 21 July 2024

 - Discussion on History of requests UI.
 - Corrections:
    - Only expand first group in history pane.
    - Move request sheet button to right side.
    - Body should show only the content type selected.
 - Discussed on auto clear history feature.
 - Auto clearing should happen at the start of app.
 - Retention periods:
    - 1 Week (default)
    - 1 Month
    - 3 Months
    - Forever

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 01 August 2024

 - UI Corrections
    - [x] Make code pane in requests similar to history.
    - [x] Move duplicate, request to inside URL Card in full screen.
 - Upcoming PRs:
    - [x] UI corrections PR.
    - [x] Env. feat tests PR.
    - [x] His. feat tests PR.
    - [x] Integration PR.
 - [ ] Figure out isolating auto-clearing.

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 08 August 2024

 - [x] PR with improvable code and sample test case with outcome desired. #442  
 - Check out spot package for integration tests.
 - [x] Create a simple app with basic implementation of hive and provider and write integration test.

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 10 August 2024

 ## Discussion on Integration tests blocker

### Issue
Integration test preview hangs at "Test starting" splash screen.

![integration_splash_screen](https://github.com/user-attachments/assets/f897d3f7-dd1b-407b-a6f1-4f80822ac659)

Attempting the test in android also had the same result
Trying to run the test with `flutter drive` command gave the following logs
```
...
VMServiceFlutterDriver: Isolate found with number: 986834591817471
VMServiceFlutterDriver: Isolate is paused at start.
VMServiceFlutterDriver: Attempting to resume isolate
...
```
Searching on the net found these issues that might be the possible reason:
- [integration_test package pauses isolates and calling pumpAndSettle() never finishes.](https://github.com/flutter/flutter/issues/73355)
- [flutter_driver pauses all isolates at their startup (even ones started from compute()), rather than only pausing initial isolate](https://github.com/flutter/flutter/issues/24703)
- [Integration test example boilerplate to support async main](https://github.com/flutter/flutter/issues/88549)
- [Can't get past splash screen after integration_test migration](https://stackoverflow.com/questions/67762969/cant-get-past-splash-screen-after-integration-test-migration)
- [Flutter Driver hangs at splash screen](https://stackoverflow.com/questions/64797858/flutter-driver-hangs-at-splash-screen)

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 14 August 2024

 - Integration test folders:
    - desktop
    - mobile
 - Integration test PRs:
    - [x] Env Manager.
    - [ ] Request Editor.
    - [x] History of requests.

**Comment by DenserMeerkat** [comment]
**Meeting Notes**
Date: 18 August 2024

 - Discussion on Integration test implementation.
 - Discussion on Final submission preparation.
 - Submission Blog.

**Comment by DenserMeerkat** [comment]
**Remaining Action items**

 - [ ] Create folder name docs and document non intuitive features.
 - [ ] Add share button in View Code page.
 - [ ] Tryout different styling for global environment sidebarcard.
 - [ ] Figure out a way to make environment suggestions work with header typeahead.
 - [ ] Figure out isolating auto-clearing.
